modules/caddyhttp/reverseproxy/upstreams.go

```diff
@@ -0,0 +1,279 @@
+package reverseproxy
+
+import (
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/caddyserver/caddy/v2"
+	"go.uber.org/zap"
+)
+
+func init() {
+	caddy.RegisterModule(SRVUpstreams{})
+	caddy.RegisterModule(AUpstreams{})
+}
+
+// SRVUpstreams provides upstreams from SRV lookups.
+// The lookup DNS name can be configured either by
+// its individual parts (that is, specifying the
+// service, protocol, and name separately) to form
+// the standard "_service._proto.name" domain, or
+// the domain can be specified directly in name by
+// leaving service and proto empty. See RFC 2782.
+//
+// Lookups are cached and refreshed at the configured
+// refresh interval.
+//
+// Returned upstreams are sorted by priority and weight.
+type SRVUpstreams struct {
+	// The service label.
+	Service string `json:"service,omitempty"`
+
+	// The protocol label; either tcp or udp.
+	Proto string `json:"proto,omitempty"`
+
+	// The name label; or, if service and proto are
+	// empty, the entire domain name to look up.
+	Name string `json:"name,omitempty"`
+
+	// The interval at which to refresh the SRV lookup.
+	// Results are cached between lookups. Default: 1m
+	Refresh caddy.Duration `json:"refresh,omitempty"`
+
+	logger *zap.Logger
+}
+
+// CaddyModule returns the Caddy module information.
+func (SRVUpstreams) CaddyModule() caddy.ModuleInfo {
+	return caddy.ModuleInfo{
+		ID:  "http.reverse_proxy.upstreams.srv",
+		New: func() caddy.Module { return new(SRVUpstreams) },
+	}
+}
+
+// String returns the RFC 2782 representation of the SRV domain.
+func (su SRVUpstreams) String() string {
+	return fmt.Sprintf("_%s._%s.%s", su.Service, su.Proto, su.Name)
+}
+
+func (su *SRVUpstreams) Provision(ctx caddy.Context) error {
+	su.logger = ctx.Logger(su)
+	if su.Proto != "tcp" && su.Proto != "udp" {
+		return fmt.Errorf("invalid proto '%s'", su.Proto)
+	}
```

Comment:
The doc on the SRVUpstreams struct says:
```
// or the domain can be specified directly in name by
// leaving service and proto empty.
```

But the implementation doesn't follow.

Reply:
I also think that's needlessly restrictive, the implementation doesn't really care what the Proto is, users might use something like `_http` instead of `_tcp` and that should be fine too :man_shrugging:

Reply:
@mohammed90 so actually `or the domain can be specified directly in name by leaving service and proto empty.` does work, that functionality is implemented in stdlib https://pkg.go.dev/net#Resolver.LookupSRV

I need to adjust how I'm doing the Caddyfile for this, I wrote this function to parse the `_<service>._<proto>` stuff out, but I need to switch it to not do that, it's not useful because just setting `name` inline should work just fine.

Reply:
> But the implementation doesn't follow.

This is done by the standard library. :+1:

Reply:
I'm also going to remove this condition here, the stdlib does not actually care what the proto is, the stdlib docs just "suggest" that `tcp` or `udp` are recommended but it's just used to assemble the name to be queried.